### PR TITLE
Fix - Trend charts look very broken with new dashboard grid changes

### LINKED ISSUE #31628: Trend charts look very broken with new dashboard grid changes
## Context
We updated the dashboard grid row height in https://github.com/metabase/metabase/pull/31019 and it's impacting Trend charts to make them look broken. See the first three cards in the top row of [this dashboard](https://stats.metabase.com/dashboard/357).

## Expected Behavior
- Update font size and spacing to look like it did before
- Make sure this is the case as small as 2x2 (the new min size). We should be able to lean on similar logic that exists for Scalars which truncates the field name with `...` when needed.

### PR DESCRIPTION
**Pull request by kamilmielnik**
Closes #31628

### Description

This fixes overflow in:
- simple scalar charts (screenshots: [before](https://github.com/metabase/metabase/assets/6830683/d5d121fb-8f0d-4bce-a4e7-955ae6545256) / [after](https://github.com/metabase/metabase/assets/6830683/1d46ff18-76d7-4666-8f41-b6cac3eb68ab))
- trend charts (screenshots: [before](https://github.com/metabase/metabase/assets/6830683/fba4ee07-a93d-4aaf-bb84-f37fe6bcfcc0) / [after](https://github.com/metabase/metabase/assets/6830683/3125ac69-78a6-4901-8014-15ca345552b2))

It works down to 2x2 size in `smartscalar`, so this PR also updates the minimum allowed size of `smartscalar`.

To see how the regression looked like please [watch these recordings](https://github.com/metabase/metabase/issues/31628#issuecomment-1597120018).

Related Slack threads:
- https://metaboat.slack.com/archives/C057T1QTB3L/p1687167736234729
- https://metaboat.slack.com/archives/C505ZNNH4/p1687463683427169

### Tests

I went with e2e tests because in unit tests elements don't have dimensions and mocking the visual state that we expect would defy the purpose.

### How to verify

Create a dashboard and add a bunch of questions visualized as: "number" and "trend" (here's the question for the trend chart I'm using in my recordings: [screenshot](https://github.com/metabase/metabase/assets/6830683/d14143a9-61c2-4b6f-8bfe-31001484f523)).

The following things affect how the cards are displayed, please try various combinations:
- card size
- viewport size (browser window size)
- whether left-hand side navigation sidebar is open or not
- whether the card has a description or not (there will be an "i" icon displayed)
- size of the text values displayed in the card (card value, card name, trend difference, etc.)

Less important but also affecting:
- font (user setting)
- language (user setting)

How it's meant to work:
- card title will be displayed in 2 lines if possible, in 1 line otherwise
- card value & title will be truncated with ellipsis if there's not enough horizontal space
- card value font size is automatically computed based on the available horizontal & vertical space (there is a min & max to this)
- if something is truncated with ellipsis, a tooltip with a complete value should be shown on hover
- if something is not truncated with ellipsis, there should be no tooltip
- description icon is shown only when description is present and description is shown in tooltip on hover
- in trend cards:
  - the `"was XYZ last month"` label will be displayed only if it fits horizontally - otherwise it will be shown in tooltip on hover
    - except on mobile resolutions, where it will be displayed in a separate row
  - the % difference will be always shown
    - it will try to use up to 2 decimal points precision if possible - otherwise it will try to use less precision
    - if there's not enough room to show it anyway - it will be truncated
    - icon is always shown

#### Trend - Before

https://github.com/metabase/metabase/assets/6830683/5ffbd7a2-f95a-4e40-90b9-79656751330b

#### Trend - After

https://github.com/metabase/metabase/assets/6830683/8ed64ce7-baaf-48a6-8274-b81114f3b41c

#### Number - Before

https://github.com/metabase/metabase/assets/6830683/e06da892-6e07-4bd4-8cb1-51d699fe43c3

#### Number - After

https://github.com/metabase/metabase/assets/6830683/0c41745f-fb86-4fb7-b193-bcd642d678ec

### CODE REVIEW

File: frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx [hunk 1]
```diff
@@ -145,17 +195,51 @@ export default class Smart extends Component {
...
+      weight: 900,
+    }).width;
+
+    const granularityDisplay = jt`last ${granularity}`;
+    const previousValueDisplay = formatValue(
+      previousValue,
+      settings.column(column),
     );
+    const disabledSeparator = ""; // avoid creating new translation key
+    const previousValueContent = jt`${disabledSeparator} was ${previousValueDisplay} ${granularityDisplay}`;
```
Comment (kamilmielnik) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx:228]: I'm not sure why separator was originally included in the translation.

Please let me know if I should introduce a new translation key.

File: e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  cy.findAllByTestId("dashcard").each(dashcard => {
+    const descendants = dashcard.find(descendantsSelector);
+
+    descendants.each((_index, descendant) => {
+      assertDescendantNotOverflowsContainer(descendant, dashcard[0]);
+    });
+  });
+};
+
+const assertDescendantNotOverflowsContainer = (descendant, container) => {
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:177]: This function is generic and could be used in other tests. Do you think I should make it a helper or a command?

**Review by cdeweyx — APPROVED** [review]

File: frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts [hunk 3]
```diff
@@ -44,14 +45,15 @@ describe("findSize", () => {
...
   it("returns the first size with which text width is smaller than the target width", () => {
     measureTextSpy
-      .mockImplementationOnce(createMockMeasureText(120))
-      .mockImplementationOnce(createMockMeasureText(110))
-      .mockImplementationOnce(createMockMeasureText(100)) // this is the one we want
-      .mockImplementationOnce(createMockMeasureText(90));
+      .mockImplementationOnce(createMockMeasureText(120, 50))
+      .mockImplementationOnce(createMockMeasureText(110, 50))
+      .mockImplementationOnce(createMockMeasureText(100, 50)) // this is the one we want
+      .mockImplementationOnce(createMockMeasureText(90, 50));
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts:51]: nit: I think to somewhat show the actual behavior you should shrink the height proportionally to the width, since `findSize` will use the same text but with reduced `size` each time. So you could get a result that will be narrower and shorter each time.

File: frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts [hunk 4]
```diff
@@ -61,11 +63,61 @@ describe("findSize", () => {
...
+
+    expect(size).toEqual("5rem");
+  });
+
+  it("returns the first size with which text height is smaller than the target height", () => {
+    measureTextSpy
+      .mockImplementationOnce(createMockMeasureText(50, 120))
+      .mockImplementationOnce(createMockMeasureText(50, 110))
+      .mockImplementationOnce(createMockMeasureText(50, 100)) // this is the one we want
+      .mockImplementationOnce(createMockMeasureText(0, 90));
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts:100]: Similiar to my previous comment, I think we should kind of proportionally, mock the width and height. They don't need to maintain the exact aspect ratio, but at least they need to shrink together.

Also, I'm not sure if the last line that sets the width to `0` is a typo or not.

**Review by WiNloSt — COMMENTED** [review]
This is more than half of the full review. I'll continue reviewing this afterward.

File: frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx [hunk 5]
```diff
@@ -23,21 +23,26 @@ interface ScalarValueWrapperProps {
 }
 
 export const ScalarValueWrapper = styled.h1<ScalarValueWrapperProps>`
+  display: inline;
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:26]: Ellipsis does not work without it

File: frontend/src/metabase/visualizations/visualizations/Scalar.jsx [hunk 6]
```diff
@@ -29,6 +29,26 @@ function legacyScalarSettingsToFormatOptions(settings) {
     .value();
 }
 
+const PADDING = 32;
+const TITLE_ICON_SIZE = 10;
+const SCALAR_TITLE_LINE_HEIGHT = 23;
+
+const TITLE_2_LINES_THRESHOLD = 120; // determined empirically
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/Scalar.jsx:36]: 823ed32d14e7232b4c2378f013f45be2aba7d790

File: frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts (hunk 4 above)
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts]: > Also, I'm not sure if the last line that sets the width to 0 is a typo or not.

Yes, good catch :+1: 
f0219dd4d85a2c4e01ac6d71bcf2ed1dd0db2505

Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ScalarValue/utils.unit.spec.ts]: c88f7c77f9b474a5c28325b3a74d314f56cc6b10

File: frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx [hunk 7]
```diff
@@ -54,14 +59,18 @@ export const ScalarTitleContent = styled.h3`
...
 export const ScalarDescriptionPlaceholder = styled.div`
   width: 1.5rem;
+  margin-top: 0.2rem;
+  padding-right: 0.5rem;
+`;
+
+export const ScalarDescriptionIcon = styled(Icon)`
+  &:hover {
+    color: ${color("brand")};
+  }
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:76]: Actually, I forgot to move `cursor: pointer`. Fixed: 27152d77b98d036fc39e237a694d091446bfb9ba

[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:76]: That's clear. Thanks.

File: frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx [hunk 8]
```diff
@@ -54,14 +59,18 @@ export const ScalarTitleContent = styled.h3`
...
-
-  &:hover {
-    color: ${color("brand")};
-  }
 `;
 
 export const ScalarDescriptionPlaceholder = styled.div`
   width: 1.5rem;
+  margin-top: 0.2rem;
+  padding-right: 0.5rem;
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:68]: There isn't an RFC about this yet, but this is a part of introducing Metabase components or what we've been doing in the project UI library. [Here's some of the conversation](https://metaboat.slack.com/archives/C057WD5L0JG/p1686227773463059?thread_ts=1686219268.256349&cid=C057WD5L0JG), and [a PR that starts using `Box`](https://github.com/metabase/metabase/pull/31426) component in our app (the description is outdated though, nothing there is about `createStyles`)

I think the thing we're trying to avoid is to stop positioning elements inside the element itself if that makes sense. Generally, I view a component as a standalone UI without either margin or padding. And the parent components that want to use any specific UI need to know where to position those UI themselves that's why I think leaving spacing e.g. `margin` and `padding` in the parent component would make more sense.

This will enable us to use reusable components with less friction as well, previously we need to create a new `*.styled.*` file if it doesn't exist. Extend the component you want to use with `styled(YourComponent)` then add one or 2 lines for the spacing. That won't be necessary with `Box` and `Flex` and all other layout components Mantine provide anymore.

[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:68]: Thank you for a detailed explanation! :100: 

If you don't mind I will skip this until we establish a convention.

[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/ScalarValue/ScalarValue.styled.tsx:68]: Sounds fine.

File: frontend/src/metabase/visualizations/visualizations/Scalar.jsx [hunk 9]
```diff
@@ -29,6 +29,27 @@ function legacyScalarSettingsToFormatOptions(settings) {
...
+const PADDING = 32;
+const TITLE_ICON_SIZE = 10;
+const SCALAR_TITLE_LINE_HEIGHT = 23;
+
+const TITLE_2_LINES_HEIGHT_THRESHOLD = 120; // determined empirically
+
+const getTitleLinesCount = height =>
+  height > TITLE_2_LINES_HEIGHT_THRESHOLD ? 2 : 1;
+
+const getTitleHeight = ({ isDashboard, showSmallTitle, titleLinesCount }) => {
```
Comment (ranquild) [inline frontend/src/metabase/visualizations/visualizations/Scalar.jsx:41]: wdyt about moving these utilities to a different (utility) file? I'd also recommend taking a look at how static viz charts (e.g. RowChart) are implemented where there are utility functions that return props for "view" components. Here we could have a function that returns values for props for inner components so there won't be any computation in the component; and we could also test these utility functions

File: frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx [hunk 10]
```diff
@@ -21,17 +21,67 @@ import {
...
+const TOOLTIP_ICON_SIZE = 11;
+const ICON_MARGIN_RIGHT = SPACING;
+const SCALAR_TITLE_LINE_HEIGHT = 23;
+const MIN_PREVIOUS_VALUE_SIZE = 27;
+const TITLE_2_LINES_HEIGHT_THRESHOLD = 150; // determined empirically
+
+const getTitleLinesCount = height =>
+  height > TITLE_2_LINES_HEIGHT_THRESHOLD ? 2 : 1;
+
+const formatChangeAutoPrecision = (
```
Comment (ranquild) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx:53]: same here

File: frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx [hunk 11]
```diff
@@ -21,17 +21,67 @@ import {
...
+  PreviousValueWrapper,
+  Separator,
   Variation,
+  VariationIcon,
+  VariationTooltip,
+  VariationValue,
 } from "./SmartScalar.styled";
 
-export default class Smart extends Component {
+const SPACING = parseInt(space(1), 10);
```
Comment (ranquild) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx:42]: I believe we shouldn't rely on `space` returning pixels. I'd rather put an explicit `8` constant here instead

**Review by ranquild — COMMENTED** [review]
Code-wise it looks like the PR does the right thing, but it adds a lot of computation logic directly to components and makes them a bit hard to follow. I believe we should follow existing conventions, e.g. ones from static-viz, and move all the computation logic to utilities in a separate file. `RowChart` could be used as an example.

We should also request for review from the team who is responsible for viz.

File: frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx (hunk 11 above)
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar.jsx]: aea584d13d4292b7a81bfc86cce3e4567603fee2

File: frontend/src/metabase/visualizations/visualizations/Scalar.jsx (hunk 9 above)
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/Scalar.jsx]: With pleasure :rocket: 

07038432ac74e5f7bad50a7e957dbfa66de149b5 + d502456f3743789e72c5179fee5896180902c5bd

Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/Scalar.jsx]: > and we could also test these utility functions

I didn't add any tests for them. There are 2 types of functions that I managed to extract:
- trivial ones involving arithmetic, which I think don't deserve tests (TS seems enough of a check)
- those using `measureText` - it'd be awkward to unit-test them, as we'd have to mock `measureText` - not sure if this whole endeavor would make sense - please let me know if you think so

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx [hunk 12]
```diff
@@ -145,17 +164,32 @@ export default class Smart extends Component {
...
+      width: getChangeWidth(width),
+    });
+
+    const granularityDisplay = jt`last ${granularity}`;
+    const previousValueDisplay = formatValue(
+      previousValue,
+      settings.column(column),
     );
+    const disabledSeparator = ""; // avoid creating new translation key
+    const previousValueContent = jt`${disabledSeparator} was ${previousValueDisplay} ${granularityDisplay}`;
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:183]: I'm not sure why we couldn't do the same as `tooltipSeparator` e.g.

```suggestion
    const previousSeparator = (
      <PreviousValueSeparator>•</PreviousValueSeparator>
    );
    const previousValueContent = jt`${previousSeparator} was ${previousValueDisplay} ${granularityDisplay}`;
```

This makes sure the value we use for both tooltip and the value (non-tooltip) uses a similar structure.

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx [hunk 13]
```diff
@@ -224,23 +262,46 @@ export default class Smart extends Component {
...
+                <PreviousValue
+                  gridSize={gridSize}
+                  id="SmartScalar-PreviousValue"
+                  responsive
+                >
+                  <PreviousValueSeparator gridSize={gridSize}>
+                    •
+                  </PreviousValueSeparator>
+
+                  {previousValueContent}
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:299]: nit: the tooltip uses
`{jt`${tooltipSeparator} was ${previousValueDisplay} ${granularityDisplay}`}`

but the non-tooltip version extracts those to a variable `previousValueContent`. I think both are similar and should use the same code style. e.g. either inlining both or extracting both to a variable.

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:1]: One thing I found confusing about this file is that there are 2 similar items 
1. The previous value in the tooltip
2. The previous value in the card
They kind of reference each other's value but they don't use the same structure when rendering. I think this could be simplified more.

**Review by WiNloSt — COMMENTED** [review]
This is 80-90% review. I'll review this the first thing tomorrow, as I can't continue now 🤯

**Review by ranquild — APPROVED** [review]
Looks good, however we need to add tests to the extracted utils. Please take a look at comments from @WiNloSt  - I agree that this can be further simplified.

File: e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+        ]);
+      });
+
+      it("should truncate value and show value tooltip on hover", () => {
+        const scalarContainer = cy.findByTestId("scalar-container");
+
+        scalarContainer.then($element => assertIsEllipsified($element[0]));
+        scalarContainer.realHover();
+
+        popover().within(() => {
```
Comment (alxnddr) [inline e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:130]: nit: you could write `popover().findByText("18,760");`

**Comment by kamilmielnik** [comment]
> Looks good, however we need to add tests to the extracted utils.

@ranquild which utils do you have in mind?

**Review by alxnddr — APPROVED** [review]
I agree with comments from Alex P and Kelvin, besides that looks good

File: e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js [hunk 15]
```diff
@@ -0,0 +1,524 @@
...
+    describe("1x2 card", () => {
+      beforeEach(() => {
+        restore();
+        cy.signInAsAdmin();
+        setupDashboardWithQuestionInCards(SCALAR_QUESTION, [
+          { size_x: 1, size_y: 2, row: 0, col: 0 },
+        ]);
+      });
+
+      it("should truncate value and show value tooltip on hover", () => {
```
Comment (WiNloSt) [inline e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:124]: Separating tests could make it easier to understand however with the application our size, this increases the test time significantly.
#### This PR which has 3 tests in `1x2 card`
Screenshot 2023-06-29 at 1 20 52 PM

#### If we assert everything in the same test
Screenshot 2023-06-29 at 1 21 55 PM

Especially when each test doesn't make any changes to the application it's safe to assert different things in the same test to put less burden on our CI runners.

Comment (kamilmielnik) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:124]: > Separating tests could make it easier to understand however with the application our size

"this becomes even more important" is another good ending to that sentence. Yeah, it's a balance.

But it feels wrong to sacrifice quality of tests because of the infrastructure. I'd rather improve the infrastructure.

----

@WiNloSt I didn't get from your comment whether you're suggesting to do this or just making me aware.
Please let me know if you think I should change it.

Comment (WiNloSt) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:124]: I was suggesting you combine all the assertions into a single test, sorry that it wasn't clear. Looping @nemanjaglumac in, since he has the most insights to our CI and he suggested me the same thing.

> But it feels wrong to sacrifice quality of tests because of the infrastructure. I'd rather improve the infrastructure.

you made a good point though.

File: e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js (hunk 14 above)
Comment (kamilmielnik) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js]: 80af793aa1d5265f5f2d698de112b1922340baa0

File: frontend/src/metabase/visualizations/visualizations/Scalar/utils.ts [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  PADDING,
+  SCALAR_TITLE_LINE_HEIGHT,
+  TITLE_2_LINES_HEIGHT_THRESHOLD,
+  TITLE_ICON_SIZE,
+} from "./constants";
+
+export const getTitleLinesCount = (height: number) =>
+  height > TITLE_2_LINES_HEIGHT_THRESHOLD ? 2 : 1;
+
+export const getTitleHeight = ({
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/Scalar/utils.ts]: 33943c19d79ae5aec031e3e02903191e492aa90a

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx [hunk 17]
```diff
@@ -224,23 +262,46 @@ export default class Smart extends Component {
...
+                </Variation>
+              </Tooltip>
+
+              {canShowPreviousValue && (
+                <PreviousValue
+                  gridSize={gridSize}
+                  id="SmartScalar-PreviousValue"
+                  responsive
+                >
+                  <PreviousValueSeparator gridSize={gridSize}>
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx]: 0f9e6015e826ad8712fbabdf288f6e76896ba3b6

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.styled.tsx [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+  color: ${color("text-light")};
+`;
+
+export const Separator = styled.span`
+  display: inline-block;
+  transform: scale(0.7);
+`;
+
+export const PreviousValueSeparator = styled(Separator)<{
+  gridSize?: { width: number };
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.styled.tsx]: 0f9e6015e826ad8712fbabdf288f6e76896ba3b6

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx (hunk 12 above)
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:183]: > why we couldn't do the same as `tooltipSeparator`

Tooltip will always show the separator.
In the card, separator will not be shown if this this text appears in a separate line, which depends on CSS breakpoints. 
Because it depends on CSS breakpoints, we can't use the same styled component in the card and in the tooltip.

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx (hunk 13 above)
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:292]: Answered here: https://github.com/metabase/metabase/pull/31722#discussion_r1245285127
Let's keep the discussion there, I will resolve this comment.

Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx]: 64d5d1a71632869e49d0b5a15b99ccc74af25130

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx (hunk 12 above)
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx]: Alright! Things are much simpler now: 84ae77da84eed1b125f48002e5a788404d673f68

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:1]: Things should be simpler now: 64d5d1a71632869e49d0b5a15b99ccc74af25130 + 84ae77da84eed1b125f48002e5a788404d673f68

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+  }).width;
+
+  const availablePreviousValueWidth =
+    getWidthWithoutSpacing(width) -
+    (changeWidth + 2 * SPACING + ICON_SIZE + ICON_MARGIN_RIGHT);
+
+  return availablePreviousValueWidth >= previousValueWidth;
+};
+
+export const concatRecursively = (value: string | string[]) => {
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts]: a379593833ad693c920858a023c5df85aa86bb24

File: frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:1]: + a379593833ad693c920858a023c5df85aa86bb24

File: e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js (hunk 15 above)
Comment (kamilmielnik) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:124]: I started doing it but it starts to look very cloudy and I can't follow it anymore:
- Even though I wrote this code, it's hard for me to say where a test begins and ends, especially if there were more than 3 `it` calls
- Variables clash - after changing viewport size we often need to reacquire reference to the DOM element and we'd have to go with `let`
- Tests are supposed to be isolated and independent - the suggestion goes against it

IMO we should not be doing this.

**Review by WiNloSt — APPROVED** [review]
Looks awesome. We can revisit the test later once Nemanja has provided some more insights for us.

File: e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js (hunk 15 above)
Comment (kamilmielnik) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:124]: > We can revisit the test later once Nemanja has provided some more insights for us.

The bad news is that e2e tests are timing out, so I might do it after all: https://github.com/metabase/metabase/actions/runs/5409602915/jobs/9830378200?pr=31722

Comment (WiNloSt) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js:124]: 🙃

Comment (kamilmielnik) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js]: Optimized and added a comment: faadfa0c10132767c887922441fe0cd0b5cb1413
I used `/** */` comments to make test cases more prominent.

Comment (WiNloSt) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js]: Now we pray to the CI god 🤞

File: frontend/src/metabase/visualizations/visualizations/Scalar.jsx (hunk 9 above)
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/visualizations/Scalar.jsx]: Added unit tests for most functions: bd8bb0e7daa8a192e0f114dde071d0c437b52bf4 + 408df899c245c0a396fcadfb8d20f7a9e0405e4d

File: e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js (hunk 15 above)
Comment (kamilmielnik) [reply e2e/test/scenarios/dashboard/reproductions/31628-responsive-scalars.cy.spec.js]: In the end I had to adjust the tests because of the optimization:
- after the change, a single test was testing 2 tooltips, so I needed to trigger `mouseleave` to make sure at most 1 tooltip is displayed at a time
- test cases with multiple `cy.viewport()` calls were flaky unless they were wrapped in `it` block, so I wrapped them back

This change made tests very difficult to debug.
I hope we will improve our infrastructure so that we don't need to compromise on quality.

File: frontend/src/metabase/visualizations/shared/components/RowChart/RowChart.tsx [hunk 20]
```diff
@@ -48,7 +48,7 @@ export interface RowChartProps<TDatum> {
 
   tickFormatters?: ChartTicksFormatters;
   labelsFormatter?: (value: NumberValue) => string;
-  measureText: TextMeasurer;
+  measureText: TextWidthMeasurer;
```
Comment (iethree) [inline frontend/src/metabase/visualizations/shared/components/RowChart/RowChart.tsx:51]: we changed the type here, but never the function that was passed, causing https://github.com/metabase/metabase/issues/32044

Comment (paoliniluis) [reply frontend/src/metabase/visualizations/shared/components/RowChart/RowChart.tsx:51]: why didn't typescript catch this?

Comment (iethree) [reply frontend/src/metabase/visualizations/shared/components/RowChart/RowChart.tsx:51]: The visualization parent component is still javascript
